lib/trino-filesystem/src/main/java/io/trino/filesystem/memory/MemoryFileSystemCache.java

```diff
@@ -0,0 +1,218 @@
+            Span span = createSpan(key, delegate.location(), "delegateLength");
+            return withTracing(span, delegate::length);
+        }
+
+        return cachedEntry.get().length();
+    }
+
+    @Override
+    public void expire(Location location)
+            throws IOException
+    {
+        List<String> expired = cache.asMap().keySet().stream()
+                .filter(key -> key.contains(location.path()))
+                .collect(toImmutableList());
+        cache.invalidateAll(expired);
+    }
+
+    @Managed
+    public void flushCache()
+    {
+        cache.invalidateAll();
+    }
+
+    @Managed
+    public long getHitCount()
+    {
+        return cache.stats().hitCount();
+    }
+
+    @Managed
+    public long getRequestCount()
+    {
+        return cache.stats().requestCount();
+    }
+
+    @Managed
+    public long getLargeFileSkippedCount()
+    {
+        return largeFileSkippedCount.get();
+    }
+
+    @VisibleForTesting
+    boolean isCached(String key)
+    {
+        Optional<Slice> cachedEntry = cache.getIfPresent(key);
+        return cachedEntry != null && cachedEntry.isPresent();
+    }
+
+    private Optional<Slice> getOrLoadFromCache(String key, TrinoInputFile delegate)
+            throws IOException
+    {
+        try {
+            return cache.get(key, () -> load(key, delegate));
+        }
+        catch (ExecutionException e) {
+            throw handleException(delegate.location(), e.getCause());
+        }
+    }
+
+    private Optional<Slice> load(String key, TrinoInputFile delegate)
+            throws IOException
+    {
+        Span span = createSpan(key, delegate.location(), "loadCache");
+        return withTracing(span, () -> {
+            long fileSize = delegate.length();
+            span.setAttribute(CACHE_FILE_READ_SIZE, fileSize);
+            if (fileSize > maxContentLengthBytes) {
+                return Optional.empty();
+            }
+            try (TrinoInput trinoInput = delegate.newInput()) {
+                return Optional.of(trinoInput.readTail(toIntExact(fileSize)));
+            }
+        });
+    }
+
+    private Span createSpan(String key, Location location, String name)
+    {
+        return tracer.spanBuilder("MemoryFileSystemCache." + name)
+                .setAttribute(CACHE_KEY, key)
+                .setAttribute(CACHE_FILE_LOCATION, location.toString())
+                .startSpan();
+    }
+
+    private static IOException handleException(Location location, Throwable cause)
+            throws IOException
+    {
+        if (cause instanceof FileNotFoundException || cause instanceof NoSuchFileException) {
+            throw withCause(new FileNotFoundException(location.toString()), cause);
+        }
+        if (cause instanceof FileAlreadyExistsException) {
+            throw withCause(new FileAlreadyExistsException(location.toString()), cause);
+        }
+        throw new IOException(cause.getMessage() + ": " + location, cause);
```

Comment:
Putting `cause.getMessage()` is redundant since you also pass `cause`

Reply:
Actually this doesn't seem to be redundant, javadoc says
```
Note that the detail message associated with cause is not automatically incorporated into this exception's detail message.
```